README.md
# OT Platform. Statistic Commands Plugin

Additional commands for statistic work with data in OTL.

## Getting Started

You need published local dispatcher-sdk lib in order not to use unmanaged libs.

### Command List
#### rare - finds the least common values in the selected fields of event.
    | rare [limit] <field-list> [by-clause]

Arguments:
- limit - the number of top rows to be output in the result. The default is 10.
  If value of this argument is 0, than all rows will be output in the result.
- field-list - comma-separated field names to which the command will apply.
- by - defines the field by which grouping occurs in the table.

#### percentile - calculates the percentile of values of defined column.
    | percentile (<field> [as <resultfield>]) [value=double] [frequency=int] 

Arguments:
- field - name of field to which the command will apply.
- value - the value of percentage. Must be between 0.0 and 1.0. Default value: 1.0.
- frequency - the value of frequency for percentile. Must be positive integer number. Default value: 1.

#### percentile_approx - calculates the approximated percentile of values of defined column.
    | percentile_approx (<field> [as <resultfield>]) [value=double] [accuracy=int]

Arguments:
- field - name of field to which the command will apply.
- value - the value of percentage. Must be between 0.0 and 1.0. Default value: 1.0.
- accuracy - the value of accuracy for approximation. Must be positive integer number. Default value: 10000.

## Running the tests

sbt test

## Deployment

See Readme.md in root of Software Development Kit project.

## Dependencies

- dispatcher-sdk_2.11  1.2.1
- sbt 1.5.8
- scala 2.11.12

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the tags on this repository.

## Authors

Dmitriy Nikolaev (dnikolaev@isgneuro.com)

## License

[OT.PLATFORM. License agreement.](LICENSE.md)

CHANGELOG.md

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2023-02-07
### Added
- percentile command
- percentile_approx command

## [1.0.4] - 2023-01-26
### Fixed
- no time to explain

## [1.0.0] - 2022-04-29
### Added
- This CHANGELOG.md.
- rare command

src/test/scala/ot/dispatcher/plugins/stat/OTLPercentileTest.scala
package ot.dispatcher.plugins.stats.commands

import org.scalatest.Matchers
import ot.dispatcher.sdk.core.{CustomException, SimpleQuery}
import ot.dispatcher.sdk.test.CommandTest

class OTLPercentileTest extends CommandTest with Matchers {

  override val dataset: String =
    """[
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"0\", \"random_Field\": \"100\", \"WordField\": \"qwe\", \"junkField\": \"q2W\", \"second_Field\": \"aa\"}","_nifi_time":"1568037188486","serialField":"0","random_Field":"100","WordField":"qwe","junkField":"q2W","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488751"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"1\", \"random_Field\": \"-90\", \"WordField\": \"rty\", \"junkField\": \"132_.\", \"second_Field\": \"bb\"}","_nifi_time":"1568037188487","serialField":"1","random_Field":"-90","WordField":"rty","junkField":"132_.","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"2\", \"random_Field\": \"50\", \"WordField\": \"uio\", \"junkField\": \"asd.cx\", \"second_Field\": \"cc\"}","_nifi_time":"1568037188487","serialField":"2","random_Field":"50","WordField":"uio","junkField":"asd.cx","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"3\", \"random_Field\": \"20\", \"WordField\": \"GreenPeace\", \"junkField\": \"XYZ\", \"second_Field\": \"aa\"}","_nifi_time":"1568037188487","serialField":"3","random_Field":"20","WordField":"GreenPeace","junkField":"XYZ","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"4\", \"random_Field\": \"20\", \"WordField\": \"fgh\", \"junkField\": \"123_ASD\", \"second_Field\": \"aa\"}","_nifi_time":"1568037188487","serialField":"4","random_Field":"20","WordField":"fgh","junkField":"123_ASD","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"5\", \"random_Field\": \"50\", \"WordField\": \"jkl\", \"junkField\": \"casd(@#)asd\", \"second_Field\": \"cc\"}","_nifi_time":"1568037188487","serialField":"5","random_Field":"50","WordField":"jkl","junkField":"casd(@#)asd","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"6\", \"random_Field\": \"60\", \"WordField\": \"zxc\", \"junkField\": \"QQQ.2\", \"second_Field\": \"aa\"}","_nifi_time":"1568037188487","serialField":"6","random_Field":"60","WordField":"zxc","junkField":"QQQ.2","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"7\", \"random_Field\": \"50\", \"WordField\": \"RUS\", \"junkField\": \"00_3\", \"second_Field\": \"cc\"}","_nifi_time":"1568037188487","serialField":"7","random_Field":"50","WordField":"RUS","junkField":"00_3","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"8\", \"random_Field\": \"0\", \"WordField\": \"MMM\", \"junkField\": \"112\", \"second_Field\": \"dd\"}","_nifi_time":"1568037188487","serialField":"8","random_Field":"0","WordField":"MMM","junkField":"112","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"10\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"9","random_Field":"10","WordField":"USA","junkField":"word","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"80\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"10","random_Field":"80","WordField":"USA","junkField":"RRR","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"70\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"11","random_Field":"70","WordField":"USA","junkField":"qwerty","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"110\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"12","random_Field":"110","WordField":"USA","junkField":"12334t","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"-40\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"13","random_Field":"-40","WordField":"USA","junkField":"r8u","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"8\", \"random_Field\": \"5\", \"WordField\": \"MMM\", \"junkField\": \"112\", \"second_Field\": \"dd\"}","_nifi_time":"1568037188487","serialField":"14","random_Field":"5","WordField":"MMM","junkField":"112","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"10\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"15","random_Field":"10","WordField":"USA","junkField":"word","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"50\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"16","random_Field":"50","WordField":"USA","junkField":"space","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"30\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"17","random_Field":"30","WordField":"USA","junkField":"two spieces","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"120\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"18","random_Field":"120","WordField":"USA","junkField":"one piece","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"},
      {"_time":1568026476854,"_meta":"_subsecond::.854 timestamp::none","host":"test.local:9990","sourcetype":"jmx","index":"test_index","source":"test_source","_raw":"{\"serialField\": \"9\", \"random_Field\": \"-50\", \"WordField\": \"USA\", \"junkField\": \"word\", \"second_Field\": \"ee\"}","_nifi_time":"1568037188487","serialField":"19","random_Field":"-50","WordField":"USA","junkField":"Amo","_subsecond":"854","timestamp":"none","_nifi_time_out":"1568037488752"}
      ]"""

  test ("Test 0. Percentile of defined column with default value (1.0). Command: | percentile random_Field") {
    val query = SimpleQuery("""random_Field""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"percentile(random_Field,1.0)":120.0}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 1. Percentile of defined column with default value and alias. Command: | percentile random_Field as rndPercentile") {
    val query = SimpleQuery("""random_Field as rndPercentile""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"rndPercentile":120.0}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 2. Error: only percentile command word. Command: | percentile") {
    val query = SimpleQuery("""""")
    the [CustomException] thrownBy{
      val command = new OTLPercentile(query, utils)
      execute(command)
    } should have message "[SearchId:-1] Required argument(s) field not found"
  }

  test("Test 3. Percentile of defined column with defined value of percentage. Command: | percentile random_Field value = 0.75") {
    val query = SimpleQuery("""random_Field value = 0.75""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"percentile(random_Field,0.75)":62.5}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 4. Percentile of defined column with defined value of percentage and alias. Command: | percentile serialField as valuedPrc value = 0.37") {
    val query = SimpleQuery("""serialField as valuedPrc value = 0.37""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"valuedPrc":7.03}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 5. Error: negative percentage. Command: | percentile random_Field value = -0.63") {
    val query = SimpleQuery("""random_Field value = -0.63""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 6. Error: percentage > 1. Command: | percentile serialField value = 6.15") {
    val query = SimpleQuery("""serialField value = 6.15""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 7. Error: text in place of percentage value. Command: | percentile random_Field value = any") {
    val query = SimpleQuery("""random_Field value = any""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 8. Percentile with default value and defined frequency. Command: | percentile random_Field frequency = 5") {
    val query = SimpleQuery("""random_Field frequency = 5""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"percentile(random_Field,1.0,5)":120.0}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 9. Percentile with defined value and defined frequency. Command: | percentile random_Field value = 0.91 frequency = 3") {
    val query = SimpleQuery("""random_Field value = 0.91 frequency = 3""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"percentile(random_Field,0.91,3)":106.90000000000005}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 10. Percentile with defined value and defined frequency, with alias. Command: | percentile random_Field as prcWithFr9 value = 0.55 frequency = 9") {
    val query = SimpleQuery("""random_Field as prcWithFr9 value = 0.55 frequency = 9""")
    val command = new OTLPercentile(query, utils)
    val actual = execute(command)
    val expected =
      """[
        |{"prcWithFr9":50.0}
        |]""".stripMargin
    assert(jsonCompare(actual, expected), f"Result : $actual\n---\nExpected : $expected")
  }

  test("Test 11. Error: negative frequency. Command: | percentile serialField value = 0.84 frequency = -6") {
    val query = SimpleQuery("""serialField value = 0.84 frequency = -6""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 12. Error: frequency = 0. Command: | percentile random_Field value = 0.23 frequency = 0") {
    val query = SimpleQuery("""random_Field value = 0.23 frequency = 0""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 13. Error: text in place of frequency. Command: | percentile random_Field frequency = newFr") {
    val query = SimpleQuery("""random_Field frequency = newFr""")
    an[IllegalArgumentException] should be thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    }
  }

  test("Test 14. Error: percentile with defined value and defined frequency, but without field defining. Command: percentile value = 0.97 frequency = 2") {
    val query = SimpleQuery("""value = 0.97 frequency = 2""")
    the[CustomException] thrownBy {
      val command = new OTLPercentile(query, utils)
      execute(command)
    } should have message "[SearchId:-1] Required argument(s) field not found"
  }

}


src/main/scala/ot/dispatcher/plugins/stats/commands/OTLRare.scala
package ot.dispatcher.plugins.stats.commands

import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions.{col, count, lit, row_number}
import ot.dispatcher.sdk.core.SimpleQuery
import ot.dispatcher.sdk.core.extensions.StringExt._
import ot.dispatcher.sdk.{PluginCommand, PluginUtils}

/** =Abstract=
 * This class provides support of __'''rare'''__ otl command.
 *
 * __'''rare'''__ finds the least common values in the selected fields of event,
 *                calculates the number and percentage of how often the selected values occur in events.
 *
 * __'''rare'''__ takes one required and two optional arguments:
 *
 * Required argument:
 *
 *    1.  _'''field-list'''_ - comma-separated field names to which the command will apply.
 *
 * Optional arguments:
 *
 *    1. __'''limit'''__ - the number of result rows to be output in the result. The default is 10.
 *    If value of this argument is 0, than all rows will be output in the result.
 *
 *    2.__'''by'''__ - defines the field by which grouping occurs in the table.
 *
 * =Usage examples=
 * OTL 1:
 * {{{| makeresults | eval a = 10,b=200 | append [makeresults | eval a =20,b=300] | append [makeresults | eval a =40,b=400]
 * | append [makeresults | eval a =30,b=400] | append [makeresults | eval a =20,b=700] | append [makeresults | eval a =10,b=300]
 * | append [makeresults | eval a =30,b=200] | append [makeresults | eval a =10,b=500] | rare a}}}
 * Result:
 * {{{+------+--------+
|    a| count| percent|
+---  +------+--------+
|   40|     1|    12.5|
|   20|     2|      25|
|   30|     2|      25|
|   10|     3|    37.5|
+-----+------+--------+}}}
 * OTL 2:
 * {{{| makeresults | eval a = 10,b=200 | append [makeresults | eval a =20,b=300] | append [makeresults | eval a =40,b=400]
 * | append [makeresults | eval a =30,b=400] | append [makeresults | eval a =20,b=700] | append [makeresults | eval a =10,b=300]
 * | append [makeresults | eval a =30,b=200] | append [makeresults | eval a =10,b=500] | rare 2 a}}}
 * Result:
 * {{{+------+--------+
|    a| count| percent|
+---  +------+--------+
|   40|     1|    12.5|
|   20|     2|      25|
+-----+------+--------+}}}
 * OTL 3:
 * {{{| makeresults | eval a = 10,b=200 | append [makeresults | eval a =20,b=300] | append [makeresults | eval a =40,b=400]
 * | append [makeresults | eval a =30,b=400] | append [makeresults | eval a =20,b=700] | append [makeresults | eval a =10,b=300]
 * | append [makeresults | eval a =30,b=200] | append [makeresults | eval a =10,b=500] | rare 0 b}}}
 * Result:
 * {{{+------+--------+
|    b| count| percent|
+---  +------+--------+
|  500|     1|    12,5|
|  700|     1|    12,5|
|  300|     2|      25|
|  200|     2|      25|
|  400|     2|      25|
+-----+------+--------+}}}
 * OTL 4:
 * {{{| makeresults | eval a = 10,b=200 | append [makeresults | eval a =20,b=300] | append [makeresults | eval a =40,b=400]
 * | append [makeresults | eval a =30,b=400] | append [makeresults | eval a =20,b=700] | append [makeresults | eval a =10,b=300]
 * | append [makeresults | eval a =30,b=200] | append [makeresults | eval a =10,b=500] | rare 3 b}}}
 * Result:
 * {{{+------+--------+
|    b| count| percent|
+---  +------+--------+
|  500|     1|    12,5|
|  700|     1|    12,5|
|  300|     2|      25|
+-----+------+--------+}}}
 * @constructor creates new instance of [[OTLRare]]
 * @param sq [[SimpleQuery]]
 */
class OTLRare(sq: SimpleQuery, utils: PluginUtils) extends PluginCommand(sq, utils, seps = Set("by")) {
  override def transform (_df: DataFrame): DataFrame = {
    val limit = args.split(" ").headOption match {
      case Some(lim) => lim.toIntSafe match {
        case Some(v) => if (v == 0) _df.count else v
        case _ => 10
      }
      case _ => return _df
    }
    val fields = returns.flatFields.filter(ff => ff.stripBackticks() != limit.toString && ff.stripBackticks() != 0.toString)
    val groups = getPositional("by") match {
      case None | Some(List()) => List()
      case Some(l) => l.map(s => s.stripBackticks())
    }
    //Dataset, grouping by rare-applying columns or by 'by'-param column + rare-applying columns with adding column of count by each group
    val dfCount = groups ++ fields match {
      case head :: tail =>
        _df.groupBy(head, tail: _*).agg(count("*").alias("count"))
      case _ => return _df
    }
    //Windowed func spec for cases of 'by'-param existing and not existing
    val w = groups match {
      case h :: t => Window.partitionBy(h, t: _*).orderBy(col("count"))
      case _ => Window.orderBy(col("count"))
    }
    //Limiting of entries: if 'by'-param exists, limiting in each group
    val dfWindowed = dfCount.withColumn("rn", row_number.over(w))
    val dfLimit = dfWindowed.filter(col("rn") <= limit)
      .drop("rn")
    //Defining of total count of entries in dataset or in each group (if 'by-param' exists) and joining limited dataset with totals-containing dataset
    val dfJoined = groups match {
      case h :: t =>
        val jdf = _df.groupBy(h, t: _*).agg(count("*").alias("total"))
        dfLimit.join(jdf, groups)
      case _ =>
        val jdf = _df.agg(count("*").alias("total"))
        dfLimit.crossJoin(jdf)
    }
    //Defining percents
    dfJoined.withColumn("percent", lit(100) * col("count") / col("total"))
      .drop("total")
  }
}

src/main/scala/ot/dispatcher/plugins/stats/parsers/ReplaceParser.scala
package ot.dispatcher.plugins.stats.parsers

import ot.dispatcher.sdk.core.{Return, ReturnField}
import ot.dispatcher.sdk.core.parsers.DefaultParser

import scala.util.matching.Regex
import ot.dispatcher.sdk.core.extensions.StringExt._

trait ReplaceParser extends DefaultParser {

  override def returnsParser: (String, Set[String]) => Return = (args: String, seps: Set[String]) => {
    val argsFiltered = excludeKeywords(excludePositionals(args, seps), keywordsParser(args))
    val splitSpaceCommaKeepQuotes: Regex = """(?:".*?"|[^,\s])+""".r
    val repl = splitSpaceCommaKeepQuotes
      .findAllIn(argsFiltered)
      .map(x => x.replace(" ", "&ph"))
      .toList
      .mkString(" ")
    val rexAsWith = """(\S+)( (as|with) (\S+))?(\s|,|$)""".r
    val fields = rexAsWith
      .findAllIn(repl)
      .matchData
      .map(x => ReturnField(Option(x.group(4)).getOrElse(x.group(1)), x.group(1)))
      .map { case ReturnField(nf, f) => ReturnField(
        nf.replace("&ph", " ").strip("'"),
        f.replace("&ph", " ").strip("'").addSurroundedBackticks)
      }
      .toList
    Return(fields)
  }
}

src/main/scala/ot/dispatcher/plugins/stats/commands/OTLPercentile.scala
package ot.dispatcher.plugins.stats.commands

import org.apache.spark.sql.DataFrame
import ot.dispatcher.sdk.{PluginCommand, PluginUtils}
import ot.dispatcher.sdk.core.SimpleQuery
import org.apache.spark.sql.functions._
import ot.dispatcher.plugins.stats.parsers.ReplaceParser
import ot.dispatcher.sdk.core.CustomException.E00012
import ot.dispatcher.sdk.core.extensions.StringExt._

/** =Abstract=
 * This class provides support of __'''percentile'''__ otl command.
 *
 * __'''percentile'''__ calculates the percentile of values of defined column.
 *
 * __'''percentile'''__ takes one required and two optional arguments:
 *
 * Required argument:
 *
 *    1.  _'''field'''_ - name of field to which the command will apply.
 *
 * Optional arguments:
 *
 *    1. __'''value'''__ - the value of percentage. Must be between 0.0 and 1.0. Default value: 1.0.
 *
 *    2. __'''frequency'''__ - the value of frequency for percentile. Must be positive integer number. Default value: 1.
 *
 * =Usage examples=
 * OTL 1:
 * {{{| makeresults | eval a = 10,b=400 | append [makeresults | eval a =40,b=600] | append [makeresults | eval a =70,b=800]
  | append [makeresults | eval a =25,b=386] | append [makeresults | eval a =11,b=895] | append [makeresults | eval a =45,b=223]
  | append [makeresults | eval a =90,b=950] | append [makeresults | eval a =60,b=787] | percentile a as a_perc}}}
 * Result:
 *  {{{+
|a_perc|
+------+
|    90|
+------+}}}
 * OTL 2:
 * {{{| makeresults | eval a = 10,b=400 | append [makeresults | eval a =40,b=600] | append [makeresults | eval a =70,b=800]
  | append [makeresults | eval a =25,b=386] | append [makeresults | eval a =11,b=895] | append [makeresults | eval a =45,b=223]
  | append [makeresults | eval a =90,b=950] | append [makeresults | eval a =60,b=787] | percentile b as b_valuedPerc value = 0.72}}}
 * Result:
 *  {{{+----------+
|     b_valuedPerc|
+-----------------+
|803.8000000000001|
+-----------------+}}}
 * OTL 3:
 * {{{| makeresults | eval a = 10,b=400 | append [makeresults | eval a =40,b=600] | append [makeresults | eval a =70,b=800]
  | append [makeresults | eval a =25,b=386] | append [makeresults | eval a =11,b=895] | append [makeresults | eval a =45,b=223]
  | append [makeresults | eval a =90,b=950] | append [makeresults | eval a =60,b=787] | percentile b as b_PercWithFreq value = 0.48 frequency = 3}}}
 * Result:
 * {{{+-----------+
|   b_PercWithFreq|
+-----------------+
|607.4799999999998|
+-----------------+}}}
 * @constructor creates new instance of [[OTLRare]]
 * @param sq [[SimpleQuery]]
 */
class OTLPercentile(sq: SimpleQuery, utils: PluginUtils) extends PluginCommand(sq, utils) with ReplaceParser {

  val flatFields: List[String] = returns.flatFields

  val flatNewFields: List[String] = returns.flatNewFields

  val field: String = if (flatFields.nonEmpty && !Array("value".addSurroundedBackticks, "frequency".addSurroundedBackticks).contains(flatFields.head) )
    flatFields.head.stripBackticks()
  else
    throw E00012(sq.searchId, "percentile", "field")

  val percentage: Double = getKeyword("value").getOrElse(1.0).toString.toDoubleSafe match {
    case Some(value) => value match {
      case p if p >= 0.0 && p <= 1.0 => p
      case _ => throw new IllegalArgumentException(s"value should be from 0.0 to 1.0, but it is $value")
    }
    case _ => throw new IllegalArgumentException(s"value should has double type, but it has other type")
  }

  val frequencyDefault: Int = 1

  val frequency: Int = getKeyword("frequency").getOrElse(frequencyDefault).toString.toIntSafe match {
    case Some(value) => value match {
      case f if f > 0 => f
      case _ => throw new IllegalArgumentException(s"frequency should be greater than 0, but it has value $value")
    }
    case _ => throw new IllegalArgumentException(s"frequency should has integer type, but it has other type")
  }

  val alias: String = createAlias()

  def transform(_df: DataFrame): DataFrame = {
    val exprText = if (frequency == frequencyDefault) {
      s"percentile($field, $percentage)"
    } else {
      s"percentile($field, $percentage, $frequency)"
    }
    val column = expr(exprText).as(alias)
    _df.agg(column)
  }

  private def createAlias(): String = {
    if (flatNewFields.nonEmpty && flatFields.nonEmpty && flatNewFields.head != flatFields.head.stripBackticks)
      flatNewFields.head
    else
      s"percentile($field,$percentage${if (frequency > frequencyDefault) "," + frequency else ""})"
  }

}